Cython/Compiler/Parsing.py

```diff
@@ -3077,18 +3087,49 @@ def p_with_gil(s):
     else:
         return 0
 
-def p_exception_value_clause(s):
+def p_exception_value_clause(s, ctx):
+    """
+    Parse exception value clause.
+
+    Maps clauses to exc_check / exc_value as follows:
+     ______________________________________________________________________
+    |                             |            |             |             |
+    | Clause                      | exc_clause | exc_check   | exc_value   |
+    | ___________________________ | __________ | ___________ | ___________ |
+    |                             |            |             |             |
+    | <nothing> (default func.)   | False      | True        | None        |
+    | <nothing> (cdef extern)     | False      | False       | None        |
+    | noexcept                    | True       | False       | None        |
+    | except <val>                | True       | False       | <val>       |
+    | except? <val>               | True       | True        | <val>       |
+    | except +                    | True       | '+'         | None        |
+    | except +*                   | True       | '+'         | '*'         |
+    | except +<PyErr>             | True       | '+'         | <PyErr>     |
+    | ___________________________ | __________ | ___________ | ___________ |
+    """
+    exc_clause = False
     exc_val = None
-    exc_check = 0
-    if s.sy == 'except':
+    if ctx.visibility  == 'extern':
+        exc_check = False
+    else:
+        exc_check = True
+
+    if s.sy == 'IDENT' and s.systring == 'noexcept':
+        exc_clause = True
+        s.next()
+        exc_check = False
+    elif s.sy == 'except':
+        exc_clause = True
         s.next()
         if s.sy == '*':
-            exc_check = 1
+            exc_check = True
             s.next()
         elif s.sy == '+':
             exc_check = '+'
             s.next()
-            if s.sy == 'IDENT':
+            if p_nogil(s):
+                ctx.nogil = 1
```

Comment:
I don't know if this is cheating, but because `except +` can be followed by an identifier (a  handler function), we have to first check if the identifier is `nogil`...